multidimensional-pod-autoscaler/AEP.md

```diff
@@ -0,0 +1,534 @@
+- Vertical and horizontal scaling executions are separated (can be avoided after in-place resizing of pod is enabled without eviction)
+- State changes in pod sizes are not persisted (too much to keep in etcd, could use Prometheus to store pod state changes)
+
+### Action Recommendation Implementation
+
+To generate the vertical scaling action recommendation, we reuse VPA libraries as much as possible to implement scaling algorithm integrated with the newly generated MPA API code.
+To do that, we need to update accordingly the code which read and update the VPA objects to be interacting with the MPA objects.
+To generate the horizontal scaling action recommendation, we reuse HPA libraries, integrating with the MPA API code, to reads and updates the MPA objects.
+We integrate vertical and horizontal scaling in a single feedback cycle.
+As an intitial solution, vertical scaling and horizontal scaling is performed alternatively (vertical scaling first).
+In the future, we can consider more complex way of prioritization and conflict resolution.
+The separation of recommendation and actuation allows customized recommender to be used to replace the default recommender.
+For example, users can plug-in their RL-based controller to replace the MPA recommender, receiving measurements from the Metrics Server and modifying the MPA objects directly to give recommendations.
+
+The implementation of the MPA framework (the backend) is based on the existing HPA and VPA codebase so that it only requires minimum code maintenance.
+Reused Codebase References:
+- HPA: https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/podautoscaler
+- VPA: https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler
+
+### MPA API Object
+
+We reuse the CR definitions from the [MultidimPodAutoscaler](https://cloud.google.com/kubernetes-engine/docs/how-to/multidimensional-pod-autoscaling) object developed by Google.
+`MultidimPodAutoscaler` is the configuration for multi-dimensional Pod autoscaling, which automatically manages Pod resources and their count based on historical and real-time resource utilization.
+MultidimPodAutoscaler has two main fields: `spec` and `status`.
+
+#### `MultidimPodAutoscalerSpec`
+
+```
+# MultidimPodAutoscalerSpec
+apiVersion: autoscaling.gke.io/v1beta1
+kind: MultidimPodAutoscaler
+metadata:
+  name: my-autoscaler
+spec:
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: my-target
+  updatePolicy:
+    updateMode: Auto
+  metrics:
+  - type: Resource
+    resource:
+    # Define the target CPU utilization request here
+    name: cpu
+    target:
+      type: Utilization
+      averageUtilization: target-cpu-util
+  constraints:
+    minReplicas: min-num-replicas
+    maxReplicas: max-num-replicas
+    applicationSLA: value  # customizable SLA for application metrics such as latency or throughput
```

Comment:
I don't understand what does SLA specification look like.

Reply:
E.g., if the SLA used is latency, then it could be something like 100ms. And the replica-set can be horizontally scaled when the measured latency is greater than 100ms, i.e., the SLA value.

Reply:
It sounds to me `applicationSLA` is only useful for a particular recommender that is SLA-based. Since we're proposing MPA API supporting various recommenders I believe the `applicationSLA` belongs as config options of a concrete recommender rather than common API.

Reply:
Is the default recommender implemented as part of this AEP going to use the `applicationSLA` field?

Reply:
The default recommender implemented in this AEP will not use the `applicationSLA` field and this field will be used by users who want to implement their own recommender. E.g., an RL/ML-based recommender can have `applicationSLA` as part of the reward/loss function and thus they can have extra constraints in addition to min/max replicas.

The `applicationSLA` field is part of `constraints` for simplicity.

Reply:
I don't understand what is type of this field. Is it a string? If so what does the string represent? Does it point to a metric? Or is it a struct? If it's a struct I think we should describe what is in the struct. Or maybe it's something else?

Reply:
It was initially designed simply to be a floating point number (most application metrics like latency and throughput are floating point numbers). Let's say an application has a latency SLA of 100ms, but the actual latency is 120.8ms; then the recommender can suggest a scale-out option. I think it should cover most use cases but I'd like to learn from you about more potential production use cases.

Reply:
> The applicationSLA field is part of constraints for simplicity.

Well, it’s going to be simple for users of the particular proprietary recommender that supports this field, rather confusing for other users that need to ignore it ;) Worse if someone else comes up with their slo-based autoscaling that requires slightly different set (or format) of constraints. I still think this field belongs to an ‘options’ map under ‘recommender’ field.

Reply:
Also, to be nit picky, I think this is more of an SLO than SLA.

Reply:
Replaced SLO with SLA. I'll leave the field as optional under 'constraints' and see if there are more use cases to support moving it under the 'recommender' field. Thanks!